### DISCUSSION TITLE
Extract project version from meson.build

### DISCUSSION THREAD

**Discussion by andy5995** [KaruroChori/vs-fltk#51] [General]
## Shell

```sh
awk '
/^\s*project\(/ {
    in_project = 1
}
in_project && /version:/ {
    match($0, /version:\s*'\''([^'\''"]+)'\'',/, arr)
    if (arr[1] != "") {
        print arr[1]
        exit
    }
}
' meson.build
```

## Python

```python
#!/usr/bin/env python3

import os
import re
import sys

# Set the path to the meson.build file
here = os.path.dirname(os.path.realpath(__file__))
meson_build_path = os.path.join(here, 'meson.build')

# Try to read and extract the version from the meson.build file
try:
    with open(meson_build_path, 'r') as f:
        version_found = False
        for line in f:
            match = re.search(r"version:\s*'([^']+)'", line)
            if match:
                print(match.group(1))
                version_found = True
                break
        if not version_found:
            raise ValueError("Version not found in meson.build")
except Exception:
    # If there's an error (e.g., file not found or version not found), exit with a non-zero error code
    sys.exit(1)
```

**Comment by KaruroChori** [comment]
The authoritative source of version is actually in `package.json`. Reading a `json` file is much more easy by comparison, and will work with the rest of my tooling. Ideally `meson.build` is just updated to match it as part of `bun run codegen`.

However, there is this problem https://github.com/mesonbuild/meson/issues/13984 which is making it extremely hard. Basically `meson rewrite` will not have the same C/C++ compiler as the rest of the commands, and instead of just changing the file as requested it will randomly complain about the C version being set.

**Reply by andy5995** [reply]
> The authoritative source of version is actually in `package.json`. Reading a `json` file is much more easy by comparison, and will work with the rest of my tooling. Ideally `meson.build` is just updated to match it as part of `bun run codegen`.

And I see now that meson can output some json:

`meson introspect --projectinfo build`

**Reply by KaruroChori** [reply]
That is actually nice. It might be helpful later on, hopefully I will not just forget about it :D.  
In any case, my last commit implemented what I was saying. So now there is just one source of information for the version and less manual steps involved.
